Mathlib/Combinatorics/Enumerative/Bell.lean

```diff
@@ -0,0 +1,217 @@
+Here, we define a refinement of these numbers, that count, for any `m : Multiset ℕ`,
+the number of partitions of a set of cardinality `m.sum` whose parts have cardinalities 
+given by `m`.
+
+The definition presents it as an integer.
+
+* `Multiset.bell`: number of partitions of a set whose parts have cardinalities a given multiset 
+
+* `Multiset.bell_zero`: `Multiset.bell 0 = 1` 
+
+* `Nat.uniformBell m n` : short name for `Multiset.bell (replicate m n)`
+
+* `Multiset.bell_mul_eq` shows that
+    `m.bell * (m.map (fun j ↦ j !)).prod * 
+      Finset.prod (m.toFinset.erase 0) (fun j ↦ (m.count j)!) = m.sum !`
+
+* `Nat.uniformBell_mul_eq`  shows that 
+    `uniformBell m n * n.factorial ^ m * m.factorial = (m * n).factorial`
+
+* `Nat.uniformBell_zero`, `Nat.uniformBell_zero'`, `Nat.uniformBell_one`, `Nat.uniformBell_one':
+compute `Nat.uniformBell` when one of the parameters is `0` or `1`
+
+* `Nat.uniformBell_succ` computes `Nat.uniformBell (m + 1) n` from `Nat.uniformBell m n`
+
+## TODO
+
+Prove that it actually counts the number of partitions as indicated.
+(When `m` contains `0`, the result requires to admit repetitions of the empty set as a part.)
+
+-/
+
+open Multiset Nat
+
+theorem Nat.choose_mul_add (m) {n : ℕ} (hn : n ≠ 0) :
+    (m * n + n).choose n = (m + 1) * (m * n + n - 1).choose (n - 1) := by 
+  rw [← mul_left_inj' (mul_ne_zero (factorial_ne_zero (m * n)) (factorial_ne_zero n))]
+  set p := n - 1 
+  have hp : n = p + 1 := (succ_pred_eq_of_ne_zero hn).symm
+  simp only [hp, add_succ_sub_one]
+  calc 
+    (m * (p + 1) + (p + 1)).choose (p + 1) * ((m * (p+1))! * (p+1)!) 
+      = (m * (p + 1) + (p + 1)).choose (p + 1) * (m * (p+1))! * (p+1)! := by ring
+    _ = (m * (p+ 1) + (p + 1))! := by rw [add_choose_mul_factorial_mul_factorial]
+    _ = ((m * (p+ 1) + p) + 1)! := by ring_nf
+    _ = ((m * (p + 1) + p) + 1) * (m * (p + 1) + p)! := by rw [factorial_succ]
+    _ = (m * (p + 1) + p)! * ((p + 1) * (m + 1)) := by ring
+    _ = ((m * (p + 1) + p).choose p * (m * (p+1))! * (p)!) * ((p + 1) * (m + 1)) := by 
+      rw [add_choose_mul_factorial_mul_factorial]
+    _ = (m * (p + 1) + p).choose p * (m * (p+1))! * (((p + 1) * (p)!) * (m + 1)) := by ring
+    _ = (m * (p + 1) + p).choose p * (m * (p+1))! * ((p + 1)! * (m + 1)) := by rw [factorial_succ]
+    _ = (m + 1) * (m * (p + 1) + p).choose p * ((m * (p + 1))! * (p + 1)!) := by ring
+  
+theorem Nat.choose_mul_right (m) {n : ℕ} (hn : n ≠ 0) :
+    (m * n).choose n = m * (m * n - 1).choose (n - 1) := by 
+  by_cases hm : m = 0 
+  · simp only [hm, zero_mul, choose_eq_zero_iff]
+    exact Nat.pos_of_ne_zero hn
+  · set p := m - 1; have hp : m = p + 1 := (succ_pred_eq_of_ne_zero hm).symm
+    simp only [hp]
+    rw [add_mul, one_mul, choose_mul_add _ hn]
+
+namespace Multiset
+
+/-- Number of partitions of a set of cardinality `m.sum` 
+whose parts have cardinalities given by `m` -/
+def bell (m : Multiset ℕ) : ℕ := 
```

Comment:
I'm now realizing that #16647 is relevant here. No action needed in this PR though.